pkg/build/strategies/onbuild/onbuild.go

```diff
@@ -71,8 +73,34 @@ func (b *OnBuild) SourceTar(config *api.Config) (io.ReadCloser, error) {
 	return b.fs.Open(tarFileName)
 }
 
+func (b *OnBuild) checkAllowedUser(config *api.Config) error {
+	if config.AllowedUIDs == nil || config.AllowedUIDs.Empty() {
+		return nil
+	}
+	user, err := b.docker.GetImageUser(config.BuilderImage)
+	if err != nil {
+		return err
+	}
+	if !userutil.IsUserAllowed(user, &config.AllowedUIDs) {
+		return errors.NewBuilderUserNotAllowedError(config.BuilderImage, false)
+	}
+	cmds, err := b.docker.GetOnBuild(config.BuilderImage)
+	if err != nil {
+		return err
+	}
+	if !userutil.IsOnbuildAllowed(cmds, &config.AllowedUIDs) {
+		return errors.NewBuilderUserNotAllowedError(config.BuilderImage, true)
+	}
+	return nil
+}
+
 // Build executes the ONBUILD kind of build
 func (b *OnBuild) Build(config *api.Config) (*api.Result, error) {
+	// Check whether image user is allowed
+	if err := b.checkAllowedUser(config); err != nil {
```

Comment:
At both here and in the Build method in sti.go, I believe you could move the checkAllowedUser function and subsequent call to the GetStrategy / GetBuilderImage path in pkg/build/strategies/strategies.go.  If you inspect the image's user there (for either sti or onbuild) and find a problem, you can exit the command sooner than at the Build step (avoiding Prepare for example), as well as address the check in one place.